src/pytorch/CFConv.cpp

```diff
@@ -0,0 +1,300 @@
+            if (weights2.size(1) != numFilters)
+                throw std::runtime_error("The size of the 2nd dimension of \"weights2\" has to be equal to the 2st dimension of \"input\"");
+
+            if (biases2.dim() != 1)
+                throw std::runtime_error("The shape of \"biases2\" has to have 1 dimension");
+            if (biases2.size(0) != numFilters)
+                throw std::runtime_error("The size of \"biases2\" has to be equal to the 2st dimension of \"input\"");
+
+            if (device.is_cpu()) {
+                impl = std::make_shared<::CpuCFConv>(numAtoms, numFilters, numGaussians, cutoff, false, gaussianWidth, activation_,
+                                                     weights1.data_ptr<float>(), biases1.data_ptr<float>(), weights2.data_ptr<float>(), biases2.data_ptr<float>());
+#ifdef ENABLE_CUDA
+            } else if (device.is_cuda()) {
+                // PyTorch allow to chose GPU with "torch.device", but it doesn't set as the default one.
+                CHECK_CUDA_RESULT(cudaSetDevice(device.index()));
+                impl = std::make_shared<::CudaCFConv>(numAtoms, numFilters, numGaussians, cutoff, false, gaussianWidth, activation_,
+                                                      weights1.data_ptr<float>(), biases1.data_ptr<float>(), weights2.data_ptr<float>(), biases2.data_ptr<float>());
+#endif
+            } else
+                throw std::runtime_error("Unsupported device");
+
+            // Create the output tensors
+            const TensorOptions options = torch::TensorOptions().device(device); // Data type of float by default
+            output = torch::empty({numAtoms, numFilters}, options);
+            inputGrad = torch::empty({numAtoms, numFilters}, options);
+            positionsGrad = torch::empty({numAtoms, 3}, options);
+
+            // cudaImpl = dynamic_cast<CudaCFConv*>(Impl.get());
+        }
+
+        if (neighbors->getCutoff() != cutoff)
+            throw std::runtime_error("The cutoff of \"neighbors\" has changed");
+
+        if (positions.size(0) != numAtoms)
+            throw std::runtime_error("The size of the 1nd dimension of \"positions\" has changed");
+        if (positions.device() != device)
+            throw std::runtime_error("The device of \"positions\" has changed");
+
+        if (input.size(0) != numAtoms)
+            throw std::runtime_error("The size of the 1nd dimension of \"input\" has changed");
+        if (input.size(1) != numFilters)
+            throw std::runtime_error("The size of the 2nd dimension of \"input\" has changed");
+        if (input.device() != device)
+            throw std::runtime_error("The device of \"input\" has changed");
+
+        // if (cudaImpl) {
+        //     const torch::cuda::CUDAStream stream = torch::cuda::getCurrentCUDAStream(tensorOptions.device().index());
+        //     cudaImpl->setStream(stream.stream());
+        // }
```

Comment:
You have several blocks that refer to `cudaImpl` and are commented out.  Is this obsolete code that should be removed?  Or is there a reason you've left it in?

Reply:
These are to support the non-default CUDA streams (like https://github.com/openmm/NNPOps/pull/37). I'll finish in another PR and it needs more changes.